src/box/memtx_space.c

```diff
@@ -1463,6 +1463,16 @@ memtx_space_finish_alter(struct space *old_space, struct space *new_space)
 	}
 }
 
+/**
+ * Invalidate the space in transaction manager.
+ */
+static void
+memtx_space_invalidate(struct space *space)
+{
+	memtx_tx_abort_all_for_ddl(in_txn());
```

Comment:
Don't we need to check `memtx_tx_manager_use_mvcc_engine`? At least for the sake of optimization.

Reply:
Added the check and wrote a comment.

Reply:
I realised we should always abort concurrent transactions. Fixed:
```diff
diff --git a/src/box/memtx_space.c b/src/box/memtx_space.c
index a4e41b362e..20064e9936 100644
--- a/src/box/memtx_space.c
+++ b/src/box/memtx_space.c
@@ -1474,14 +1474,13 @@ static void
 memtx_space_invalidate(struct space *space)
 {
 	/*
-	 * Abort all concurrent transactions and invalidate space in MVCC
-	 * only if it is enabled. Otherwise, all concurrent transactions
-	 * will be aborted when DDL will be prepared.
+	 * Invalidate space in MVCC only if it is enabled. However, we should
+	 * always abort all concurrent transactions: for example, we should
+	 * abort all readers of new schema when DDL failed to commit.
 	 */
-	if (memtx_tx_manager_use_mvcc_engine) {
-		memtx_tx_abort_all_for_ddl(in_txn());
+	memtx_tx_abort_all_for_ddl(in_txn());
+	if (memtx_tx_manager_use_mvcc_engine)
 		memtx_tx_invalidate_space(space, in_txn());
-	}
 }
 
 /* }}} DDL */

```

Reply:
Nah, forget it - transactions cannot yield without MVCC. Sorry for the rush - I revert the changes.